controllers/workspace/status.go

```diff
@@ -47,6 +47,31 @@ func initCurrentStatus() currentStatus {
 	}
 }
 
+// initStoppingStatus initialize the status for stopping workspace
+// it may be STOPPING or FAILED in case it's stopped automatically due workspace startup failure
+// in the second case the corresponding condition is propagated
+func initStoppingStatus(dwStatus dw.DevWorkspaceStatus) currentStatus {
+	stoppingStatus := currentStatus{workspaceConditions: workspaceConditions{
+		conditions: map[dw.DevWorkspaceConditionType]dw.DevWorkspaceCondition{},
+	}}
+
+	if dwStatus.Phase == dw.DevWorkspaceStatusFailed {
+		stoppingStatus.phase = dw.DevWorkspaceStatusFailed
+		stoppingStatus.copyConditionFromDevWorkspace(dw.DevWorkspaceFailedStart, dwStatus)
+	} else {
+		stoppingStatus.phase = dw.DevWorkspaceStatusStopping
+	}
+	return stoppingStatus
+}
+
+// changeToStoppedIfNeeded changes status to stopped if needed
+// we keep workspace status failed if it's stopped by workspace startup failure
+func changeToStoppedIfNeeded(c *currentStatus) {
```

Comment:
I don't think this function should exist. It's confusing and assumes that the currentStatus is filled when most of the time it is not -- making use of this value would require you to first explicitly set `currentStatus.Phase = Failed`, which is only done in one part of the code. 

Also the naming isn't clear (what makes changing to stopped needed?) without reading the function doc.

Reply:
Tina sorry me for proposing these changes without thinking enough )
Proposing alternative in https://github.com/devfile/devworkspace-operator/pull/375#discussion_r613487876 thread